File: airbyte-commons-worker/src/main/java/io/airbyte/workers/models/StateWithId.java
/*
 * Copyright (c) 2020-2025 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.workers.models;

import io.airbyte.protocol.models.AirbyteMessage;
import io.airbyte.protocol.models.AirbyteStateMessage;
import java.util.OptionalInt;
import java.util.concurrent.atomic.AtomicInteger;

public class StateWithId {

  private static final String ID = "id";

  public static AirbyteMessage attachIdToStateMessageFromSource(final AirbyteMessage message) {
    if (message.getType() == AirbyteMessage.Type.STATE) {
      message.getState().setAdditionalProperty(ID, StateIdProvider.getNextId());
    }
    return message;
  }

  public static AirbyteStateMessage attachIdToStateMessageFromSource(final AirbyteStateMessage message) {
    message.setAdditionalProperty(ID, StateIdProvider.getNextId());
    return message;
  }

  public static OptionalInt getIdFromStateMessage(final AirbyteMessage message) {
    if (message.getType() == AirbyteMessage.Type.STATE) {
      return OptionalInt.of(getIdFromStateMessage(message.getState()));
    }
    return OptionalInt.empty();
  }

  public static int getIdFromStateMessage(final AirbyteStateMessage message) {
    return (int) message.getAdditionalProperties().get(ID);
  }

  private static class StateIdProvider {

    private static final AtomicInteger id = new AtomicInteger(0);

    public static int getNextId() {
      return id.incrementAndGet();
    }

  }

}

File: airbyte-container-orchestrator/src/main/kotlin/io/airbyte/container/orchestrator/worker/model/StateWithId.kt
/*
 * Copyright (c) 2020-2025 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.container.orchestrator.worker.model

import io.airbyte.protocol.models.v0.AirbyteMessage
import io.airbyte.protocol.models.v0.AirbyteStateMessage
import java.util.concurrent.atomic.AtomicInteger

const val ID = "id"

fun attachIdToStateMessageFromSource(message: AirbyteMessage): AirbyteMessage {
  if (message.type == AirbyteMessage.Type.STATE) {
    attachIdToStateMessageFromSource(message.state)
  }

  return message
}

fun attachIdToStateMessageFromSource(message: AirbyteStateMessage): AirbyteStateMessage =
  with(message) {
    if (!message.additionalProperties.contains(ID)) {
      setAdditionalProperty(ID, StateIdProvider.nextId)
    }
    return this
  }

fun getIdFromStateMessage(message: AirbyteMessage): Int? {
  if (message.type == AirbyteMessage.Type.STATE) {
    return getIdFromStateMessage(message.state)
  }
  return null
}

fun getIdFromStateMessage(message: AirbyteStateMessage): Int =
  if (message.additionalProperties.contains(
      ID,
    )
  ) {
    message.additionalProperties[ID] as Int
  } else {
    throw IllegalStateException("State message does not contain id")
  }

private object StateIdProvider {
  private val id = AtomicInteger(0)

  val nextId: Int
    get() = id.incrementAndGet()
}
